controllers/upgrade/upgradeExecutor.go

```diff
@@ -0,0 +1,165 @@
+	"context"
+	"fmt"
+
+	"github.com/cloudogu/cesapp-lib/core"
+	"github.com/cloudogu/cesapp-lib/registry"
+
+	k8sv1 "github.com/cloudogu/k8s-dogu-operator/api/v1"
+	"github.com/cloudogu/k8s-dogu-operator/controllers/cesregistry"
+	"github.com/cloudogu/k8s-dogu-operator/controllers/limit"
+	"github.com/cloudogu/k8s-dogu-operator/controllers/resource"
+
+	imagev1 "github.com/google/go-containerregistry/pkg/v1"
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type imageRegistry interface {
+	// PullImageConfig pulls a given container image by name.
+	PullImageConfig(ctx context.Context, image string) (*imagev1.ConfigFile, error)
+}
+
+type fileExtractor interface {
+	// ExtractK8sResourcesFromContainer copies a file from stdout into map of strings
+	ExtractK8sResourcesFromContainer(ctx context.Context, doguResource *k8sv1.Dogu, dogu *core.Dogu) (map[string]string, error)
+}
+
+type serviceAccountCreator interface {
+	// CreateAll creates K8s services accounts for a dogu
+	CreateAll(ctx context.Context, namespace string, dogu *core.Dogu) error
+}
+
+type doguRegistrator interface {
+	// RegisterDoguVersion registers a certain dogu in a CES instance.
+	RegisterDoguVersion(dogu *core.Dogu) error
+}
+
+type collectApplier interface {
+	// CollectApply applies the given resources to the K8s cluster but filters and collects deployments.
+	CollectApply(ctx context.Context, customK8sResources map[string]string, doguResource *k8sv1.Dogu) (*appsv1.Deployment, error)
+}
+
+type resourceUpserter interface {
+	// ApplyDoguResource generates K8s resources from a given dogu and creates/updates them in the cluster.
+	ApplyDoguResource(ctx context.Context, doguResource *k8sv1.Dogu, dogu *core.Dogu, image *imagev1.ConfigFile, customDeployment *appsv1.Deployment) error
+}
+
+type upgradeExecutor struct {
+	client                client.Client
+	imageRegistry         imageRegistry
+	collectApplier        collectApplier
+	fileExtractor         fileExtractor
+	serviceAccountCreator serviceAccountCreator
+	doguRegistrator       doguRegistrator
+	resourceUpserter      resourceUpserter
+}
+
+func NewUpgradeExecutor(client client.Client, imageRegistry imageRegistry, collectApplier collectApplier, fileExtractor fileExtractor, serviceAccountCreator serviceAccountCreator, registry registry.Registry) *upgradeExecutor {
+
+	doguRegistrator := cesregistry.NewCESDoguRegistrator(client, registry, nil)
+	limitPatcher := limit.NewDoguDeploymentLimitPatcher(registry)
+	upserter := resource.NewUpserter(client, limitPatcher)
+
+	return &upgradeExecutor{
+		client:                client,
+		imageRegistry:         imageRegistry,
+		collectApplier:        collectApplier,
+		fileExtractor:         fileExtractor,
+		serviceAccountCreator: serviceAccountCreator,
+		doguRegistrator:       doguRegistrator,
+		resourceUpserter:      upserter,
+	}
+}
+
+func (ue *upgradeExecutor) Upgrade(ctx context.Context, toDoguResource *k8sv1.Dogu, toDogu *core.Dogu) error {
```

Comment:
Missing docs comment for exported function

- [ ] Done

Reply:
Also I would like to see more general events to inform the user about the current upgrade process à la:

```
... Registering upgraded version <version> in local dogu registry
... Pulling new image registry.cloudogu.com/cloudogu/ldap:<version>
... (Applying custom resources in cluster)
... Updating dogu resources in cluster
```

- [ ] Done

Reply:
Yes that makes sense to have a better overview about the current actions